### PR TITLE
Vishal/unstaked an 1

### PR DESCRIPTION
**Pull request by vishalchangrani**
Refactored the code to address this [comment](https://github.com/onflow/flow-go/pull/911#issuecomment-879261062).

Now, the structure is as follows:
1. A new interface [NodeBuilder](https://github.com/onflow/flow-go/blob/vishal/unstaked_an/cmd/node_builder.go) defines all methods needed to bootstrap a Flow node.
2. Existing [FlowNodeBuilder](https://github.com/onflow/flow-go/blob/vishal/unstaked_an/cmd/scaffold.go#L97) and the new [AccessNodeBuilder](https://github.com/onflow/flow-go/blob/vishal/unstaked_an/cmd/access/access_node_builder.go#L29) implement this interface.
3. A new interface [AccessNodeBuilder](https://github.com/onflow/flow-go/blob/vishal/unstaked_an/cmd/access/access_node_builder.go#L21) which extends the NodeBuilder interface and adds Access Node specific functions.
4. Separate [StakedAccessNodeBuilder](https://github.com/onflow/flow-go/blob/vishal/unstaked_an/cmd/access/staked_access_node_builder.go) and [UnstakedAccessNodeBuilder](https://github.com/onflow/flow-go/blob/vishal/unstaked_an/cmd/access/unstaked_access_node_builder.go) (which are composed of `AccessNodeBuilder`) to take care of bringing up Staked and Unstaked AN resp.
5. Changes to all existing startup files (`main.go`) to use the `NodeBuilder` functions instead of directly using the implementation.

### CODE REVIEW

File: cmd/access/main.go
Comment (vishalchangrani) [inline cmd/access/main.go:116]: At this point, we know what type of access node this is (staked or unstaked) and we then proceed with the appropriate node builder.

File: cmd/access/main.go
Comment (vishalchangrani) [inline cmd/access/main.go:129]: All calls to the base node builder is via the new interface `NodeBuilder`

File: cmd/access/staked_access_node_builder.go
Comment (vishalchangrani) [inline cmd/access/staked_access_node_builder.go:11]: wanted to avoid as much as possible the `if staked then else ` paradigm hence added two new sub node builder sub classes.

File: cmd/collection/main.go
Comment (vishalchangrani) [inline cmd/collection/main.go:130]: All `main.go` now use the interface `cmd.NodeBuilder` instead of the actual implementation `FlowNodeBuilder`

File: cmd/node_builder.go
Comment (vishalchangrani) [inline cmd/node_builder.go:23]: The new interface

File: network/p2p/network.go
Comment (vishalchangrani) [inline network/p2p/network.go:23]: to address @smnzhu comment earlier for making this a constant

**Review by huitseeker — COMMENTED** [review]
So, to start: the `NodeBuilder` interface is great.

I have more questions about the `AccessNodeBuilder` interface. Here's the core idea:
In the end, we will have not 2 but 3 types of access nodes (Unstaked AN, Staked AN participating on the unstaked network, Staked AN not participating on the unstaked network).

The question is: how easy is it at runtime (from CLI options) to compose a Staked AN participating on the Unstaked net from:
- the bootstrapping code from the Staked AN that does not participate in the unstaked network,
- the bootstrapping code from the Unstaked AN.

(note that when I say "compose" in the above, I'm talking about mixing concrete components, not composing sub-interfaces, which might only have one implementation anyway)

I also expect that a lot of the components currently indifferentiated in the access' main (e.g. node clients, follower engine, ...) will end up in only one of these bootstrapers depending on the above use cases. We do not have to effect the migration of these components in this PR, but it would be good if we could point to logical places where they would live in the future.

File: cmd/access/access_node_builder.go
Comment (vishalchangrani) [inline cmd/access/access_node_builder.go:52]: these are 3 util methods that can be used by staked or unstaked AN to create an unstaked network

File: cmd/access/staked_access_node_builder.go
Comment (huitseeker) [inline cmd/access/staked_access_node_builder.go:80]: Nit: technically, we would initialize with those other staked ANs that participate in the public network. That's probably better left to a follow-up PR, but a TODO comment here would be nice.

File: cmd/access/access_node_builder.go
Comment (huitseeker) [reply cmd/access/access_node_builder.go:52]: Nit: Those  methods are great, but would deserve a comment per, to describe the span of their work (some of these comments are in `enqueueUnstakedNetworkInit`).

File: cmd/access/staked_access_node_builder.go
Comment (huitseeker) [inline cmd/access/staked_access_node_builder.go:61]: Nit: Because the public identity (at least key, and maybe ID as well) should not be the same, we'll want to leave a TODO here.

**Review by huitseeker — COMMENTED** [review]
Thanks for your latest changes, which I think will make this much easier to test. How about we merge this in #911 and may see from if / how we want to break up the code changes?
(I let a few nits)

**Review by huitseeker — APPROVED** [review]

File: cmd/access/main.go
Comment (smnzhu) [inline cmd/access/main.go:132]: Why this change?

File: cmd/access/staked_access_node_builder.go
Comment (smnzhu) [reply cmd/access/staked_access_node_builder.go:80]: @vishalchangrani The plan is to eventually have more than one staked AN that participates in the public network, right?

File: cmd/node_builder.go
Comment (smnzhu) [reply cmd/node_builder.go:103]: > I wonder if it would be possible to put these into a separate data struct, rather than together with the interface, and have all implementations embed both?

Another possibility is to have a single method that returns some `NodeData` struct, which contains all of these fields. This would still not be entirely ideal, but might be an easier change to make for now?

I think eventually, it might make sense to separate the `NodeBuilder` interface (the functionality) from the data entirely, having a separate struct encapsulate the latter. The methods `Module`, `Component`, `PostInit`, etc. could then be passed in `f func(builder NodeBuilder, data NodeData) error` instead of `f func(builder NodeBuilder) error`.

This can probably be in a separate PR though.

File: cmd/access/unstaked_access_node_builder.go
Comment (huitseeker) [reply cmd/access/unstaked_access_node_builder.go:61]: It's really a shim at the moment, since the public NodeIDs (incl keys) should ultimately be found in the bootstrap files, but I agree a TODO comment marking how ephemeral the situation is would help.

File: cmd/access/unstaked_access_node_builder.go
Comment (huitseeker) [reply cmd/access/unstaked_access_node_builder.go:103]: See also [this comment](https://github.com/onflow/flow-go/pull/983#discussion_r672335745)

File: cmd/node_builder.go
Comment (huitseeker) [reply cmd/node_builder.go:103]: While finessing the getter/setter approach may be a bit premature for now, I think:
- it's [ok](https://golang.org/doc/effective_go#Getters) if not quite idiomatic,
- we're defining the interface at a time where we do have numerous instances for it, which is the one thing Go does recommend,

so I'm tempted to defer to @vishalchangrani 's good sense. I would point out though that this is already a big PR, and probably daunting to other reviewers => any change that reduces the # of SLoC, even at some marginal cost in flexibility of this interface, hastens the moment of merge.

File: cmd/access/access_node_builder.go
Comment (vishalchangrani) [reply cmd/access/access_node_builder.go:69]: yup - added comments to each of those 3 method

File: cmd/access/staked_access_node_builder.go
Comment (vishalchangrani) [reply cmd/access/staked_access_node_builder.go:85]: eventually, yes. But initially I would like to just keep it to one staked AN which could support upto 10 to 15 unstaked AN.
Adding more than one ANs to the unstaked network would mean that all those AN also see unnecessary network traffic.

File: cmd/access/access_node_builder.go
Comment (jordanschalm) [inline cmd/access/access_node_builder.go:31]: Thanks for the diagram 🎨

**Review by jordanschalm — APPROVED** [review]
👏 Nice work

File: cmd/node_builder.go
Comment (vishalchangrani) [inline cmd/node_builder.go:105]: @huitseeker @smnzhu @jordanschalm Added this new struct to hold all the data for the node. The `NodeBuilder` functions modifies this struct through its methods.

File: cmd/node_builder.go
Comment (vishalchangrani) [reply cmd/node_builder.go]: Thanks for that excellent suggestion @smnzhu. I did make that change and now have a `NodeConfig` struct which holds the data that is modified (removing all the Get/Set from the `NodeBuilder` interface).
The change actually reduces the number of lines changed since the existing code uses `node.dot.xxx` everywhere and now is now just `NodeConfig`

File: cmd/access/staked_access_node_builder.go
Comment (vishalchangrani) [reply cmd/access/staked_access_node_builder.go]: yup - made that change.

**Review by smnzhu — APPROVED** [review]
Looks good :)

**Pull request merged**
